고양이 장난감가게 유효성 추가하기, 회원 관리 만들기

eunhasoo opened the pull request:
안녕하세요! 어제는 일이 생겨서 작업을 못했습니다 🥲
상품 도메인은 커밋되어 있는 환경 그대로 이용했고,

- [x] 회원 도메인 추가
- [x] 회원 생성, 수정, 삭제 API 개발
- [x] 회원 Service, Controller 테스트

작업했습니다! 리팩토링과 리뷰해주시는 부분들 반영해보겠습니다.
이번주도 잘 부탁드립니다 :)

eunhasoo committed:
build: Dozer Mapper 의존성 추가, 쿼리 로그 출력 설정

eunhasoo committed:
chore: 불필요한 코드 정리

eunhasoo committed:
refactor: 상품에 적용된 lombok 코드 수정

- AllArgsConstructor 제거
- 생성자 만들고 @Builder 적용

eunhasoo committed:
feat: 회원 도메인 추가

eunhasoo committed:
test: 회원 생성 테스트 추가

eunhasoo committed:
feat: UserRepository, UserService 추가

eunhasoo committed:
feat: 회원 정보 수정하는 update 메소드 추가

eunhasoo committed:
refactor: 회원 정보 삭제시 삭제된 회원 정보를 반환하도록 수정

eunhasoo committed:
test: UserService 테스트 추가

eunhasoo committed:
feat: UserService 구현

- DozerMapper 빈 등록
- 생성/수정을 위한 DTO 생성
- 회원 조회 실패시 던질 예외 생성

eunhasoo committed:
fix: 잘못 import된 라이브러리 수정

- javax -> springframework

eunhasoo committed:
test: UserController 테스트 추가

eunhasoo committed:
feat: UserController 구현

- 요청 DTO 입력 검증 추가
- 응답 DTO 추가 (password 제외)

eunhasoo committed:
feat: UserNotFoundException 예외시 응답 추가

- @RestControllerAdvice로 변경
- @CrossOrigin 적용

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    @Override
+    public User update(Long id, UserUpdateRequest source) {
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException(id + "에 해당하는 회원을 찾지 못해 수정할 수 없습니다."));
+
+        user.update(source.getName(),
+                source.getEmail(),
+                source.getPassword());
+
+        return user;
```
Optional을 사용하시는군요! 
Optional의 기능을 좀 더 사용해서 `user` 라는 변수를 없애보면 좋은 경험이 될 거에요!

eunhasoo replied:
넵 수정해보겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/UserUpdateRequest.java:
```diff
@@ -0,0 +1,30 @@
+    private String password;
+
+    @Builder
+    @ConstructorProperties({"name", "email", "password"})
```
이 어노테이션을 어떤 일을 하나요?!

eunhasoo replied:
Jackson 라이브러리는 JSON을 자바 객체로 변환할 때 기본 생성자를 기반으로 생성합니다. 기본 생성자가 없으면 에러가 발생하는데, 기본 생성자를 만들고 싶지 않을 때 따로 정의한 생성자에 `@ConstructorProperties`를 붙여주면 동일한 이름을 가진 JSON 속성값을 생성자에 넘겨줍니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceImplTest.java:
```diff
@@ -0,0 +1,169 @@
+        @Test
+        @DisplayName("회원을 생성하여 반환한다")
+        void it_returns_created_user() {
+            User user = userService.create(subject());
```
subject 안에 `create` 도 있어야겠어요! 테스트하는 것이  `subject` 니까요!

eunhasoo committed:
refactor: update 메소드 user 변수 제거

eunhasoo committed:
test: subject 메소드 내부로 create 호출 이동

eunhasoo committed:
refactor: 불필요한 클래스 제거

eunhasoo committed:
refactor: 패키지 구조 정리

- Exception 한곳에서 관리하도록 이동
- ProductData -> ProductRequest 네이밍 변경
- ProductRepository 하나로 사용

eunhasoo committed:
feat: 입력 검증 실패 메시지 추가

- ProductRequest 클래스에 id 속성 삭제

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/UserServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    @Override
+    public User delete(Long id) {
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException(id + "에 해당하는 회원을 찾지 못해 삭제할 수 없습니다."));
+
+        userRepository.delete(user);
+
+        return user;
```
이것도 위 메서드처럼 가능하겠는데요?!

eunhasoo replied:
방법을 찾아보는데 delete 메소드에 리턴값이 없어서 변수를 없애기가 어려울 것 같습니다 ㅠㅠ

ssisksl77 replied:
```suggestion
        return userRepository.findById(id)
                .map(user -> {
                    userRepository.delete(user);
                    return user;
                }).orElseThrow(() -> new UserNotFoundException(id + "에 해당하는 회원을 찾지 못해 삭제할 수 없습니다."));
```

이 방법은 어떨까요?

eunhasoo replied:
오 이렇게 하면 될 것 같아요..!! java 8 문법을 더 공부해야겠네요 😅 반영할게요 감사합니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/ProductRequest.java:
```diff
@@ -5,21 +5,19 @@
+public class ProductRequest {
 
-    @NotBlank
+    @NotBlank(message = "이름은 필수값 입니다.")
```
`MessageSource` 를 이용해보면 좋은 경험이 될거에요!

eunhasoo committed:
refactor: 입력값 유효성 검사 메시지 한곳에서 관리

eunhasoo committed:
refactor: ControllerAdvice 패키지 및 이름 변경

eunhasoo committed:
feat: 검증 실패시 사용할 응답 데이터 추가

eunhasoo committed:
refactor: 변수 정의없이 리턴값 반환하도록 수정

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceImplTest.java:
```diff
@@ -0,0 +1,172 @@
+                .build();
+    }
+
+    @Nested
+    @DisplayName("create 메소드는")
+    class Describe_create {
+        private User subject() {
+            return userService.create(
+                    UserCreateRequest.builder()
+                        .name(NAME)
+                        .email(EMAIL)
+                        .password(PASSWORD)
+                        .build()
+            );
+        }
+
+        @Test
+        @DisplayName("회원을 생성하여 반환한다")
+        void it_returns_created_user() {
+            User user = subject();
+
+            assertThat(user).isNotNull();
+            assertThat(user.getId()).isNotNull();
+        }
```
Context 가 보이지 않네요! 간단하게 나눠보는 것도 좋아보입니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/UserCreateRequest.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+public class UserCreateRequest {
+
+    @Mapping("name")
```
매핑기능을 사용하면서 느끼신 장단점을 공유해주세요!
